File: Leetcode July Challenge/Python/Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    def lengthOfLinkedList(self, head):
        length = 0
        temp = head
        while temp != None:
            temp = temp.next
            length += 1
        
        return length
            
    
    def reverseKGroupH(self, head, k, length):
        if length < k:
            return head
        
        count, nex, prev, curr = 0, None, None, head
        while count < k and curr != None:
            nex = curr.next
            curr.next = prev
            prev = curr
            curr = nex
            count += 1
        
        if nex != None:
            head.next = self.reverseKGroupH(nex, k, length - k)
        
        return prev
    
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        
        length = self.lengthOfLinkedList(head)
        return self.reverseKGroupH(head, k, length)
        
        
        

File: Leetcode July Challenge/C++/Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    int lengthOfLinkedList(ListNode* head){
        int length = 0;
        ListNode* temp = head;
        while(temp != NULL){
            temp = temp->next;
            length++;
        }
        return length;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k, int length){
        if(length < k){
            return head;
        }
        
        int count = 0;
        ListNode* prev = NULL, *nex = NULL, *curr = head;
        while(count < k && curr != NULL){
            nex = curr -> next;
            curr -> next = prev;
            prev = curr;
            curr = nex;
            count++;
        }
        
        if(nex != NULL){
            head -> next = reverseKGroup(nex, k, length - k);
        }
        
        return prev;
        
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        int length = lengthOfLinkedList(head);
        return reverseKGroup(head, k, length);
    }
};
